asciidocapi.py

```diff
@@ -177,28 +177,42 @@ def __init__(self, asciidoc_py=None):
         self.options = Options()
         self.attributes = {}
         self.messages = []
-        # Search for the asciidoc command file.
-        # Try ASCIIDOC_PY environment variable first.
+        # Search for the asciidoc command file in that order :
+        # - ASCIIDOC_PY environment variable
+        # - asciidoc_py function argument
+        # - sibling (preferred to shell search paths, to ensure version matching)
+        # - shell search paths
+        # - current working directory (may be different from the directory of this file)
+        fnames = ['asciidoc.py', 'asciidoc.pyc', 'asciidoc']
         cmd = os.environ.get('ASCIIDOC_PY')
+        is_cmd_valid = lambda c: c and os.path.isfile(c)
         if cmd:
             if not os.path.isfile(cmd):
                 raise AsciiDocError('missing ASCIIDOC_PY file: %s' % cmd)
-        elif asciidoc_py:
+
+        if not is_cmd_valid(cmd) and asciidoc_py:
             # Next try path specified by caller.
             cmd = asciidoc_py
             if not os.path.isfile(cmd):
                 raise AsciiDocError('missing file: %s' % cmd)
-        else:
+
+        if not is_cmd_valid(cmd):
+            # try to find sibling
+            THIS_PATH = os.path.dirname(os.path.realpath(__file__))
+            candidates = (find_in_path(fname, THIS_PATH) for fname in fnames)
+            cmd = next((fname for fname in candidates if is_cmd_valid(fname)), None)
+
+        if not is_cmd_valid(cmd):
             # Try shell search paths.
-            for fname in ['asciidoc.py','asciidoc.pyc','asciidoc']:
-                cmd = find_in_path(fname)
-                if cmd: break
-            else:
-                # Finally try current working directory.
-                for cmd in ['asciidoc.py','asciidoc.pyc','asciidoc']:
-                    if os.path.isfile(cmd): break
-                else:
-                    raise AsciiDocError('failed to locate asciidoc')
+            candidates = (find_in_path(fname) for fname in fnames)
+            cmd = next((fname for fname in candidates if is_cmd_valid(fname)), None)
```

Comment:
This is not an improvement either, leave it as the simple version.

Reply:
I think it is better to call an instruction defined, tested and maintained by python than doing a for/else that is error prone.
But, as usual with refactor, its a matter of taste.